src/Juvix/Core/IR/Typechecker/Types.hs

```diff
@@ -0,0 +1,182 @@
+  | ShouldBeFunctionType
+      (IR.Value' ext primTy primVal)
+      (IR.Term'  ext primTy primVal)
+  | UnboundIndex Natural
+  | SigmaMustBeZero
+  | UsageMustBeZero
+  | UsageNotCompatible Usage.T Usage.T
+  | UnboundBinder Natural IR.Name
+  | MustBeFunction
+      (IR.Elim' ext primTy primVal)
+      Natural
+      (IR.Term' ext primTy primVal)
+  | BoundVariableCannotBeInferred
+
+type TypecheckError = TypecheckError' IR.NoExt
+
+deriving instance
+  (Eq primTy, Eq primVal,
+   IR.TermAll    Eq ext primTy primVal,
+   IR.ElimAll    Eq ext primTy primVal,
+   IR.ValueAll   Eq ext primTy primVal,
+   IR.NeutralAll Eq ext primTy primVal) =>
+  Eq (TypecheckError' ext primTy primVal)
+
+instance (Show primTy, Show primVal,
+          IR.TermAll    Show ext primTy primVal,
+          IR.ElimAll    Show ext primTy primVal,
+          IR.ValueAll   Show ext primTy primVal,
+          IR.NeutralAll Show ext primTy primVal) =>
+  Show (TypecheckError' ext primTy primVal)
+ where
+  show (TypeMismatch binder term expectedT gotT) =
+    "Type mismatched. \n" <> show term <> " \n (binder number " <> show binder
+      <> ") is of type \n"
+      <> show gotT
+      <> ".\nBut the expected type is "
+      <> show expectedT <> "."
+  show (UniverseMismatch i j) =
+    "The universe " <> show i <>
+    " should be strictly less than " <> show j <> "."
+  show (CannotApply f x) =
+    "Application (vapp) error. Cannot apply \n" <> show f <> "\n to \n" <> show x
+  show (ShouldBeStar ty) =
+    "* n is of type * but " <> show ty <> " is not *."
+  show (ShouldBeFunctionType ty f) =
+    show ty <> " is not a function type but should be - while checking " <> show f
+  show (UnboundIndex n) =
+    "unbound index " <> show n
+  show (SigmaMustBeZero) =
+    "Sigma has to be 0."
+  show (UsageMustBeZero) =
+    "Usage has to be 0."
+  show (UsageNotCompatible expectedU gotU) =
+    "The usage of "
+      <> (show gotU)
+      <> " is not compatible with "
+      <> (show expectedU)
+  show (UnboundBinder ii x) =
+    "Cannot find the type of \n"
+      <> show x
+      <> "\n (binder number "
+      <> show ii
+      <> ") in the environment."
+  show (MustBeFunction m ii n) =
+    ( show m <> "\n (binder number " <> show ii
+        <> ") is not a function type and thus \n"
+        <> show n
+        <> "\n cannot be applied to it."
+    )
+  show (BoundVariableCannotBeInferred) =
+    "Bound variable cannot be inferred"
+
+type HasThrowTC' ext primTy primVal m =
+  HasThrow "typecheckError" (TypecheckError' ext primTy primVal) m
+
+type HasThrowTC primTy primVal m =
+    HasThrowTC' IR.NoExt primTy primVal m
+
+throwTC :: HasThrowTC' ext primTy primVal m
+        => TypecheckError' ext primTy primVal -> m z
+throwTC = throw @"typecheckError"
+
+
+data T
+
+do
+  let typed = Ext.Ann $ \primTy primVal -> [t|Annotation $primTy $primVal|]
+  extT <- IR.extendTerm "Term" [] [t|T|] $
+    IR.defaultExtTerm
+      { IR.typeStar   = typed
```

Comment:
`HasThrowTC'` is an abbreviation for a `HasThrow "typecheckError"` constraint, so it's just applying `throw` to the type argument to disambiguate.

The independent `do` is a TH splice. They can be arbitrary expressions, not just function calls, and I needed to share the `typed` variable between `extendTerm` and `extendElim`. Do you think it would it be more obvious that it's a splice if it was `$(do ...)` instead?